subsys/bluetooth/host/att.c

```diff
@@ -605,7 +605,55 @@ static u8_t find_type_cb(const struct bt_gatt_attr *attr, void *user_data)
 	}
 
 	/* Check if data matches */
-	if (read != data->value_len || memcmp(data->value, uuid, read)) {
+	if (read != data->value_len) {
+		/* Use bt_uuid_cmp() to compare UUIDs of different form. */
+		struct bt_uuid_128 ref_uuid;
+		struct bt_uuid_128 recvd_uuid;
+		u32_t val32;
+		u16_t val16;
+
+		/* Copy from packet data to struct bt_uuid format */
+		switch (data->value_len) {
+		case 16:
+			recvd_uuid.uuid.type = BT_UUID_TYPE_128;
+			memcpy(recvd_uuid.val, data->value, 16);
+			break;
+		case 4:
+			recvd_uuid.uuid.type = BT_UUID_TYPE_32;
+			val32 = sys_get_le32(data->value);
+			((struct bt_uuid_32 *)&recvd_uuid)->val = val32;
+			break;
+		case 2:
+			recvd_uuid.uuid.type = BT_UUID_TYPE_16;
+			val16 = sys_get_le16(data->value);
+			((struct bt_uuid_16 *)&recvd_uuid)->val = val16;
+			break;
+		}
+
+		/* Copy from internal list to struct bt_uuid format */
+		switch (read) {
+		case 16:
+			ref_uuid.uuid.type = BT_UUID_TYPE_128;
+			memcpy(ref_uuid.val, uuid, read);
+			break;
+		case 4:
+			ref_uuid.uuid.type = BT_UUID_TYPE_32;
+			val32 = *((u32_t *)uuid);
+			((struct bt_uuid_32 *)&ref_uuid)->val = val32;
+			break;
+		case 2:
+			ref_uuid.uuid.type = BT_UUID_TYPE_16;
+			val16 = *((u16_t *)uuid);
+			((struct bt_uuid_16 *)&ref_uuid)->val = val16;
+			break;
+		}
```

Comment:
While this seems correct I wonder if it wouldn't be better to create a helper function to convert from over the air format to struct bt_uuid, i.e. bt_uuid_create(&uuid, size, data), which basically consolidates the code above which is already duplicated in a few places include uuid_create that that uses a net_buf.

Reply:
I agree. I will work on a change for that.
The 32-bit UUID is not allowed over the air, so should I remove that case?
Core5.1: Vol3, partF, sect3: "All 32-bit Attribute UUIDs shall be converted to 128-bit UUIDs when the Attribute UUID is contained in an ATT PDU."

Reply:
> I agree. I will work on a change for that.
> The 32-bit UUID is not allowed over the air, so should I remove that case?
> Core5.1: Vol3, partF, sect3: "All 32-bit Attribute UUIDs shall be converted to 128-bit UUIDs when the Attribute UUID is contained in an ATT PDU."

Id leave 32bit since the uuid.c has already handling for it, we can have a check before calling the function to check if the size is valid for the bearer, or alternatively we a le variant, bt_uuid_le_create which does the check before calling bt_uuid_create.

Reply:
Is there a bug in uuid_create()? It uses net_buf_pull_le16() for the 16-bit UUID but memcpy instead of net_buf_pull_mem() for the 128-bit UUID.
If it is okay NOT to decrease buf->len and increase buf->data, in uuid_create(), then we can call the new bt_uuid_create, otherwise we can not benefit from reusing code (other than make it easier to read).

Reply:
You can always do it advance the buf data manually, regarding the reason we use memcpy is that net_buf_pull_mem don't really copy anything to bt_uuid_128 to be able to do things like bt_uuid_cmp, it would perhaps be possible to just reference with pointer instead of having val[16] but we would probably add 4 bytes extra per UUID 128 bits so I rather leave like it is and do the memcpy.